src/core/storage/ipfs-uploader.ts

```diff
@@ -0,0 +1,144 @@
+import { UploadProgressEvent } from "../../types/events";
+
+export class IpfsUploader implements IStorageUpload {
+  /**
+   * Fetches a one-time-use upload token that can used to upload
+   * a file to storage.
+   *
+   * @returns - The one time use token that can be passed to the Pinata API.
+   */
+  public async getUploadToken(contractAddress: string): Promise<string> {
+    const headers = {
+      "X-App-Name": `CONSOLE-TS-SDK-${contractAddress}`,
+    };
+    const res = await fetch(`${TW_IPFS_SERVER_URL}/grant`, {
+      method: "GET",
+      headers,
+    });
+    if (!res.ok) {
+      throw new FetchError(`Failed to get upload token`);
+    }
+    const body = await res.text();
+    return body;
+  }
+
+  public async uploadBatchWithCid(
+    files: (string | FileOrBuffer)[],
+    fileStartNumber = 0,
+    contractAddress?: string,
+    signerAddress?: string,
+    options?: {
+      onProgress: (event: UploadProgressEvent) => void;
+    },
+  ): Promise<CidWithFileName> {
+    const token = await this.getUploadToken(contractAddress || "");
+    const metadata = {
+      name: `CONSOLE-TS-SDK-${contractAddress}`,
+      keyvalues: {
+        sdk: "typescript",
+        contractAddress,
+        signerAddress,
+      },
+    };
+    const data = new FormData();
+    const fileNames: string[] = [];
+    files.forEach((file, i) => {
+      let fileName = "";
+      let fileData = file;
+      // if it is a file, we passthrough the file extensions,
+      // if it is a buffer or string, the filename would be fileStartNumber + index
+      // if it is a buffer or string with names, the filename would be the name
+      if (file instanceof File) {
+        let extensions = "";
+        if (file.name) {
+          const extensionStartIndex = file.name.lastIndexOf(".");
+          if (extensionStartIndex > -1) {
+            extensions = file.name.substring(extensionStartIndex);
+          }
+        }
+        fileName = `${i + fileStartNumber}${extensions}`;
+      } else if (file instanceof Buffer || typeof file === "string") {
+        fileName = `${i + fileStartNumber}`;
+      } else if (file && file.name && file?.data) {
+        fileData = file?.data;
+        fileName = `${file.name}`;
+      } else {
+        // default behavior
+        fileName = `${i + fileStartNumber}`;
+      }
+
+      const filepath = `files/${fileName}`;
+      if (fileNames.indexOf(fileName) > -1) {
+        throw new DuplicateFileNameError(fileName);
+      }
+      fileNames.push(fileName);
+      if (typeof window === "undefined") {
+        data.append("file", fileData as any, { filepath } as any);
+      } else {
+        // browser does blob things, filepath is parsed differently on browser vs node.
+        // pls pinata?
+        data.append("file", new Blob([fileData as any]), filepath);
+      }
+    });
+
+    data.append("pinataMetadata", JSON.stringify(metadata));
+
+    if (typeof window === "undefined") {
```

Comment:
man i would reaaaaally love this for the thirdweb CLI - what would it take to make it happen?